### DISCUSSION TITLE
uhppoted-mqtt

### DISCUSSION THREAD

**Discussion by liviom** [uhppoted/uhppoted#12] [Q&A] [answered]
Dear, 

first of all thanks for this great work.

I want to publish mqtt messages to a mosquitto browser

1) what is the right configuration for #mqtt section in uhppoted.conf?
i tried to set mosquitto address and port but receive only  these messages in uhppoted/gateway/system

_{
  "message": {
    "signature": "IwzDTi8R298RX6ToyMHSEWMgVVzscJ7KRVxi8dvhz2XqvvlX9L7fCINqyMw1yZCQ7WDqLS1lXHDHYwnJbdfGBeZhiENFWXeqN7JDwhG9QxAsZiV8ARzOIMwfxWNeFtXJy3VCSkjGhAwhHIE7OWpwPyYfN3IO97XENIMHaXYPgeLwD56IfszAVbbgUvuqj4gxwqbOrZsFPojQxrjzJc9tqO20AGVBdjZR3EGPnQXShY971iJuOXYzZUOFvjY+eQU+kG8abf2oEhPDZpFkqDob9DrwtS+/1GzampoIU/OFNJukTgJD0f44eMw5MVypfDHmR9GeJ7dVuGrFIUS03EYfeA==",
    "key": "fcfAGtbeMQz+9QU2aKUWHpcdp/T6Vt05AYdHQ5Mf6c7QHW/dlUIyjHp/TkTX6YDlk55gwoYK1CvUwxI+8aAwmG2p6TatM4UMIVhb8kRj40VT57csUVFysk/XVrp0BcXOJlVtkM8S9R0x8vfy43w2437cjQdt7Zp4AnQYfDZofSiXeHvrl3b7GpxmOSiJOvrn85gytuolGS2zz+V9R1P7tl3wdPCGumv+CO88NZHNKSn8zbFDWPHx4nzFLCnoMbIAaPkB3Y9uv26daoOH+JxTRhWnWp687wGM8kwZLKBApT5e7Rk5s8ZOnCRHCdUg8P5JcQePy3RKDkoTmGLrpqsy0Q==",
    "system": "iTVfBcwyl25KxqL5HQGixMlyCaW/0znW0mlOKumfLMJO64pvgYTr3iugvRIpaRtqfsXm8Aqt3gXt4sSOSSZibZUTI4sHwmbiBiEqOfnxdYo="
  },
  "hmac": "2a8510e6ee1e6b4d95e40d742324c326e836448b28c43c6629e37a41acf6f589"
}_

my goal is to receive in mosquitto live messages on swipe and send command for open relays

2) to reach my goal I have to run some others command (for example uhppoted-cli or others) to listen for device swipe events and then publish them to mosquitto?

thanks for the support

Livio

**Comment by uhppoted** [comment]
Very happy to hear somebody is using the MQTT stuff :-) :-) ... it always seemed like a such a flexible and robust way to integrate systems.

Ok, it looks like you have full encryption enabled - that is only necessary if you're using a public MQTT broker (e.g. the [HiveMQ dev instance](https://www.hivemq.com/public-mqtt-broker)) and are (rightly) concerned about other users being able to look at your messages.

If you're running your own broker on your own network you don't need any of that and it makes things unnecessarily complicated so start by disabling all the authentication and encryption in _uhppoted.conf_:
```
# MQTT
...
mqtt.security.authentication = NONE
mqtt.security.HMAC.required = false
mqtt.security.nonce.required = false
mqtt.security.outgoing.sign = false
mqtt.security.outgoing.encrypt = false
...
```

With those settings you should be able see plaintext messages on the mosquito client:
```
...
{
  "message": {
    "system": {
      "alert": {
        "message": "UTC0311-L0x 405419896  system time not synchronized:2022-07-05 16:33:39 (1h0m1s)",
        "subsystem": "health-check"
      }
    }
  },
  "hmac": "b2f5310884b860cd09cc50ed1ccb33bbc562d211e4fd3a118a13cd2d0408984c"
}
{
  "message": {
    "event": {
      "event": {
        "access-granted": false,
        "card-number": 8165538,
        "device-id": 405419896,
        "direction": 1,
        "door-id": 3,
        "event-id": 70,
        "event-reason": 6,
        "event-type": 1,
        "timestamp": "2022-07-05 16:34:14 PDT"
      }
    }
  },
  "hmac": "2e9255a55a3538299809257b5423fe7c6e7afb7e409658bf87497604ee1072d3"
}
...
```

I'm using _HiveMQ_ but it should be the same for _mosquito_ - if you're seeing the encrypted messages then you're probably mostly configured correctly already. Once you've got that all working we can step through tightening up the security bit-by-bit if you need to.

Notes:
1. You don't need to do anything special to receive events - they're published on the _uhppoted/gateway/events_ topic so if you listen to the _uhppoted/# topic_ you should see all the messages.
2. The MQTT stuff is still using MQTT v4 because it was written before the Paho client had support for MQTT v5 and the request/response protocol. Updating it is on the TODO list but not in the very near future.

**Comment by liviom** [comment]
dear thanks for your quick reply

this is my .conf file
[uhppoted.conf.txt](https://github.com/uhppoted/uhppoted/files/9051913/uhppoted.conf.txt)


I followed this  guide to install the latest release 0.8.0

https://github.com/uhppoted/uhppoted/discussions/11


but I only receive this message

_{"message":{"system":{"alive":{"message":"1 error","subsystem":"health-check"}}},"hmac":"15fe3b46284d111df88bd5c28af7657171e994a9944618510cb5d42383ef620b"}_

and sometimes this message

_{
  "message": {
    "system": {
      "alert": {
        "message": "UTC0311-L0x 423133701  incorrect listener address/port: 192.168.0.70:60001",
        "subsystem": "health-check"
      }
    }
  },
  "hmac": "1ee1d8cbe12c36d89533499370c52b0f8e2943364b27b41c5ea92b1305202815"
}_

no swipe record received...

where do it take the listner address?
what is wrong in my configuration?
192.168.0.70 is not the address of the ubuntu system where I lunch uhppoted-mqtt, it's the ip address of the pc wjere there is the Control Access software included in the cdrom arrived with board.

if you need I can activate a temporary ssh access from external to allow you to connect to my system, I'll send you address,port and credential in a private mail.


with the command sudo uhppote-cli --debug get-devicesthe output is 

![2022-07-06 07_49_27-livio@UbuntuOH_ ~](https://user-images.githubusercontent.com/13430820/177478278-8d920cdc-d326-4212-91f0-ed34e8f488f3.png)


192.168.0.101 is the ip of uhppote board

thanks for your support

Livio

**Comment by uhppoted** [comment]
Ok, that's looking almost good - your controller is configured to send events to the _AccessControl_ application so all you need to do is change the controller _listener_ address using _uhppote-cli_ [set-listener](https://github.com/uhppoted/uhppote-cli#set-listener) command.

For example, if the IP address of the Ubuntu system is 192.168.0.100, then:
```
uhppote-cli set-listener 423133701 192.168.0.100:60001

uhppote-cli get-listener 423133701
```

Notes:
1. By default _uhppoted-mqtt_ listens on UDP port 60001 - you can change that in the _uhppoted.conf_ file if necessary.
2. The _AccessControl_ application may change the controller listener address every time you run it - if so, so you will need to set it again after running the AccessControl application. You could just set up a cron job to do that but if you want to get more sophisticated it can be done with an MQTT message too (e.g. you could figure out some kind of alerting system using the health-check messages).
3. If you want to see _door open_ and _door close_ events you will have enable [record-special-events](https://github.com/uhppoted/uhppote-cli#record-special-events)

**Reply by liviom** [reply]
GREAT,  GREAT, GREAT

it works with the following .conf file (I upload it so it can be useful for other people)
[uhppoted.conf.txt](https://github.com/uhppoted/uhppoted/files/9056938/uhppoted.conf.txt)

but I have  to 
a) stop nodered examples that were running in background (I realize that it could be a problem after you explain me that also AccessControl software can change the listener so I immediately think about node red examples I installed)
b) reboot ubuntu  to have it working, it's normal?

so there are some questions now....
1) the events message in mosquitto will return the following information

_      "event": {
        "access-granted": true,
        "card-number": xxxxxxxx,
        "device-id": 423133701,
        "direction": 1,
        "door-id": 1,
        "event-id": 210,
        "event-reason": 1,
        "event-type": 1,
        "timestamp": "2022-07-06 19:30:41 CEST"
      }_

where I can find the code "readable" description to show in openhab home automation system? if I read event-type 1 for me is the same as event-type 1000, It could be useful have a transcode of the numeric value such as event-type, event-reason,event-id and so on, the map of numeric value-->description will be done in openhab...it's more simple for my knowns....

2) where I can find a list of accepted command and, if possible, some examples? 
you say that  "You could just set up a cron job to do that but if you want to get more sophisticated it can be done with an MQTT message too" but what is the command?

it would be great if it's possible to configure via mqtt all the function/commands that there are in node red examples (set cards, users, grant permission, start sequential learning of person.....

3) there is no way to have 2 listners to run all both AccessControl software and uhppoted-mqtt (please note that access control is running on a windows OS and uhppoted-mqtt on a linux os)???? 
where I can find the instructions to run uhppote-mqtt on windows? there is a guide for windows users? 
I know windows is not linux but I think that if I run the software  AccessControl (that is windows executable) and windows version of uhppoted-mqtt  on the same machine I can not works on a solution to continuos set and get listener, it's correct?



thanks again for your great works and support and please forgive me for my not perfect english


Livio

**Comment by uhppoted** [comment]
Awesome .. that's really great:-) :-) ..

re. _reboot Ubuntu_. That shouldn't have been necessary - best guess is that something else was already running in the background and the reboot killed it, but TBH it could have been anything.

re. _events message_. I've created an [issue](https://github.com/uhppoted/uhppoted-mqtt/issues/2) to add human readable descriptions to the event message. It's a little complicated because the text has to be reasonably easy to internationalise so it's going to take a bit of experimenting and thought. For the moment though, the easiest place to get the lookup codes is either the NodeRED project:
- [lookup.js](https://github.com/uhppoted/node-red-contrib-uhppoted/blob/master/nodes/lookup.js)

or the NodeJS project:
- [lookup.js](https://github.com/uhppoted/uhppoted-nodejs/blob/288adae5d2ab8831a8d286dca18d470b9a1558cd/src/lookup.js)
- [en-US.js](https://github.com/uhppoted/uhppoted-nodejs/blob/288adae5d2ab8831a8d286dca18d470b9a1558cd/src/locales/en-US/en-US.js)

re. _MQTT commands_. Hmmm, thought I had documented them but clearly it slipped through the cracks :-(. Ok, will add that to the ever-growing TODO list but for now the [Makefile](https://github.com/uhppoted/uhppoted-mqtt/blob/master/Makefile) has an example of every supported command message.

re. _multiple event listeners_. I **knew** that question was coming :-> !! At the moment it's not really possible because the UDP event messages need to be redistributed to multiple listeners and that needs some low level network stuff which is not cross-platform compatible. It's also a little tricky because the UHPPOTE _AccessControl_ Windows application doesn't really 'play nice' with other software. **BUT** .. the next release of the [uhppoted-tunnel](https://github.com/uhppoted/uhppoted-tunnel) is planned to include an _event connector_ which should (hopefully) implement event redistribution to multiple listeners. In the interim, you may be able to repurpose one of the other UDP tunnel type projects out there.

re. _uhppoted-mqtt on Windows_. Running it on Windows is the basically same as running it on Linux - except that if it's running as a service then you have to use the _Windows Service Manager_ instead of _systemctl_. It shouldn't conflict with the UHPPOTE AccessControl application, except that only one of them will receive the event messages.

If you can let me know once you've got your _OpenHAB_ implementation working I'll add a link to it on the main README. Am sure other people will find it useful!

**Comment by liviom** [comment]
dear,

thanks for all the answers,

Ill' check the file you link to verify the informations (mqtt commands and code description) needed.

but I try to run all both uhppoted-mqtt (on ubuntu 192.168.0.90) and Accesscontrol software (on windows server  2019 ip 192.168.0.70) and it seems work great without any mqtt-command or cron job to set and get the listener;

it is fortune or what?

attached the image: 
1. mqtt explorer for system health, uhppoted1.png
2. mqtt explorer for events, uhppoted0.png
3. AcessControl software running in "download and monitoring" mode, uhppoted2.png
4. sql server logs showing the same events , uhppoted3.png


![uhppoted3](https://user-images.githubusercontent.com/13430820/177724319-747aef9b-d3a2-460b-8cfc-7c82ae290fb2.png)
![uhppoted0](https://user-images.githubusercontent.com/13430820/177724326-bde43b67-18c5-4672-88f4-a5be6a62670a.png)
![uhppoted1](https://user-images.githubusercontent.com/13430820/177724330-3c852c96-9693-445d-b383-739f39408231.png)
![uhppoted2](https://user-images.githubusercontent.com/13430820/177724332-c57065b7-ffc6-4320-9eef-ea457ae1350e.png)

please note that there are a 1 number of difference between the ID returned from uhppoted-mqtt and the one registered from AccessControl software,

the id returned from uhppoted-mqtt is 217, the one registered from AccessControl is 216, but is the same also for the mini webserver activated, see the following image uhppoted4.png, why there is this difference?
![uhppoted4](https://user-images.githubusercontent.com/13430820/177725950-27be80d8-faa1-42d5-be00-d9a07e61fd11.png)

thanks again for your answers

**Reply by liviom** [reply]
Another question

from Makefile of uhppoted-mqtt

open-door:
	_mqtt publish --topic 'uhppoted/gateway/requests/device/door/lock:open' \
               --message '{ "message": { "request": { "request-id":  "$(REQUESTID)", \
                                                      "client-id":   "$(CLIENTID)",  \
                                                      "reply-to":    "$(REPLYTO)",   \
                                                      "device-id":   $(SERIALNO),    \
                                                      "card-number": $(CARD),        \
                                                      "door":        3 }}}'_

what are the '$(REQUESTID)  and   $(REPLYTO)' ? how to obtain them?


Just to inform you, and it's OFF TOPIC, I just created openHAB things (only for reading at the moment) and item files, I'm working on map to transform numeric values into human readable events....and after I will work on add /remove cards and opening door based on permission

**Reply by liviom** [reply]
to this I answer by myself...it could be useful for other people

REQUESTID is a generic id
reply to is the topic in wich will be published the response


this is the commnad topi where I sent the request and the payload:

topic:    uhppoted/gateway/requests/device:get
payload:
{ "message": 
  { "request": 
    { "request-id": "1234", 
      "client-id":  "clientidlivio",
      "reply-to":   "uhppoted/reply/replytolivio",
      "device-id":  423133701 
     }
  }
}



and the response arrived at topic: uhppoted/reply/replytolivio
with the following payload:
{
  "message": {
    "reply": {
      "client-id": "clientidlivio",
      "method": "get-device",
      "nonce": 1,
      "request-id": "1234",
      "response": {
        "address": {
          "IP": "192.168.0.101",
          "Port": 60000,
          "Zone": ""
        },
        "date": "2015-12-15",
        "device-id": 423133701,
        "device-type": "UTO311-L04",
        "gateway-address": "192.168.0.1",
        "ip-address": "192.168.0.101",
        "mac-address": "00:xx:xx:xx:xx:xx",
        "subnet-mask": "255.255.0.0",
        "timezone": {},
        "version": "0662"
      },
      "server-id": "uhppoted"
    }
  },
  "hmac": "87221903bc41597a4bfc232cff66991d1e127cc56fefb30c0edc1addb6f9d451"
}



it was at the top of Makefile

**Comment by uhppoted** [comment, marked as answer]
> but I try to run all both uhppoted-mqtt (on ubuntu 192.168.0.90) and Accesscontrol software (on windows server 2019 ip 192.168.0.70) and it seems work great without any mqtt-command or cron job to set and get the listener;

That's good news! It's possible the _AccessControl_ application only initialises the controller listener once (on the initial configuration).

> the id returned from uhppoted-mqtt is 217, the one registered from AccessControl is 216

_uhppoted-mqtt_ is (or at least should be) using the event index as stored on the controller, which starts at 1 so it's possible that the _AccessControl_ application is using an internal index that starts at zero. But just to verify, can you please run the following commands:
```
uhppote-cli get-event 423133701 216
uhppote-cli get-event 423133701 217
```
and see which event matches the _uhppoted-mqtt_ and _AccessControl_ logs respectively.

> what are the '(REPLYTO)' ? how to obtain them?
>   > REQUESTID is a generic id
>   > reply to is the topic in wich will be published the response

tl;dr: Basically yes :-). Somewhat longer explanation:

MQTTv3 which is the protocol used by _uhppoted-mqtt_ does not support a request-response pattern (MQTTv5 does but was not implemented by _paho_ at the time) so a request ID allows a client to correlate a response message with a request and the _reply-to_ topic allows the client to match received messages to handlers in much the same way a REST server matches URLs to handlers.

Both are optional and for client convenience only (i.e. neither are used internally in the _uhppoted-mqtt_ server):
- the _request-id_ field (if used) should be a unique identifier for every request (typically a monotonically increasing atomic integer). 
- the response message will be sent to the default _reply_ topic (`uhppoted/gateway/replies`) if no _reply-to_ field is supplied

**Comment by liviom** [comment]
Dear after updating to v.0.8.8 i receive a lot of mqtt messages from all the input, also from not linked imput of the board
how can i solve?

**Comment by uhppoted** [comment]
Hi,

There were only two changes from v0.8.7 to v0.8.8:
- The Go version was updated to v1.22
- The `restore-default-parameters` function was added

Neither of those should cause a problem - what version were you running previously? 

And can you post a sample of the messages?

**Reply by liviom** [reply]
Dear
first of all thanks for your quick reply...
the previous version was 0.8.5

`{
  "message": {
    "event": {
      "event": {
        "access-granted": false,
        "card-number": 85,
        "device-id": 423133701,
        "direction": 1,
        "direction-text": "in",
        "door-id": 4,
        "event-id": 5512,
        "event-reason": 38,
        "event-reason-text": "forced open",
        "event-type": 3,
        "event-type-text": "alarm",
        "timestamp": "2024-03-30 11:37:21 CET"
      }
    }
  },
  "hmac": "5908277d8288c9f73e3865a142a64ee301f935b55ebb6e23579e195ff38bbabf"
}`

there are a lot of this messages, about 15 every minutes.
the card 85 doesn't exists.

this is the result of uhppote-cli get-devices

` 423133701  192.168.0.101  255.255.255.0  192.168.0.1  00:66:19:38:82:05  v6.62  2015-12-15
`

but i noticed that if i start mqtt daemon it doesn't sent strange messages until an event occur, for example an opening...after that it start to send a  lot of messages.

furthermore is there a procedure to upgrade to the new firmware? where I can download it?

**Comment by uhppoted** [comment]
Ah, ok, that explains it - the events handling was reworked in [v0.8.7](https://github.com/uhppoted/uhppoted-mqtt/releases/tag/v0.8.7). There's an explanation in the [README](https://github.com/uhppoted/uhppoted-mqtt#events) with more detail and some history in these two issues if you're interested:
- [Rework events handling](https://github.com/uhppoted/uhppoted-mqtt/issues/16)
- [Old event published on each card swipe](https://github.com/uhppoted/uhppoted-mqtt/issues/15)

But - to summarise - as of v0.8.7 _uhppoted-mqtt_ has two events feeds:
- a _live_ events feed which is events as they occur
- a _historical_ events feed which is a best attempt to publish all the events stored on a controller (e.g. for logging  or storing to a database). 

The historical events feed (which is what you're seeing) will slowly retrieve **all** the events from a controller in small batches and your controller has 5500+ events which it is trying ever so desperately to catch up with, so it's going to take a while. If you just leave it running it will eventually have retrieved all the events and the feed will go back to normal.

If you only want to see _current_ events, then you need to subscribe to the _./events/live_ topic (or alternatively reconfigure the topics in your _uhppoted.conf_ file).

However, you may also want to check your controller - that _forced open_ alarm event is possibly something worth following up and if 5500+ events is more than you expect then you may have a door contact that is 'chattering'.

**re. firmware**. Unfortunately the manufacturer has stated categorically that the controller firmware cannot be upgraded. According to their [Amazon page](https://www.amazon.com/UHPPOTE-Professional-Wiegand-Network-Software):
>
> **Q: How do I upgrade the firmware on the boards?**\
> A: I am sorry the firmware can't been upgraded. thanks.\
> _By Xiaojia Huang on September 9, 2019_
> 
> **Q: Is there a way to update the firmware or software**\
> A: the software have not updated sevral year, so don't need see more\
> _By Xiaojia Huang on April 14, 2017_
> 
> **Q: Are you able to upgrade the firmware on this ? if check is made on the door using the software? how can you do this as i want to buy more**\
> A: Here are the anwsers of your questions:\
> 1\)The firmware can't be upgraded, and you could download the latest software.\
> 2\) what kind of the check do you want to made on the door using the software. We recommend that you could tell us. Thanks.\
> _By Xiaojia Huang on January 18, 2022_
> 
> 
>

**Reply by liviom** [reply]
Thanks a lot fir your answer...
